pyg_lib/csrc/sampler/cpu/subgraph_kernel.cpp

```diff
@@ -3,96 +3,139 @@
+        // need to find `col.size(0) / nodes.size(0)` neighbors on average).
+        // TODO Benchmark this customization
+        grain_size = std::max<int64_t>(out_col.size(0) / nodes.size(0), 1);
+        grain_size = at::internal::GRAIN_SIZE / grain_size;
+        at::parallel_for(
+            0, nodes.size(0), grain_size, [&](int64_t _s, int64_t _e) {
+              for (scalar_t i = _s; i < _e; ++i) {
+                const auto v = nodes_data[i];
+                // Iterate over all neighbors and check if they
+                // are part of `nodes`:
+                scalar_t offset = out_rowptr_data[i];
+                for (scalar_t j = rowptr_data[v]; j < rowptr_data[v + 1]; ++j) {
+                  const auto w = mapper.map(col_data[j]);
+                  if (w >= 0) {
+                    out_col_data[offset] = w;
+                    if (return_edge_id)
+                      out_edge_id_data[offset] = j;
+                    offset++;
+                  }
+                }
+              }
+            });
+      });
 
   return std::make_tuple(out_rowptr, out_col, out_edge_id);
 }
 
+std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>>
+subgraph_bipartite_kernel(const at::Tensor& rowptr,
+                          const at::Tensor& col,
+                          const at::Tensor& src_nodes,
+                          const at::Tensor& dst_nodes,
+                          const bool return_edge_id) {
+  TORCH_CHECK(rowptr.is_cpu(), "'rowptr' must be a CPU tensor");
+  TORCH_CHECK(col.is_cpu(), "'col' must be a CPU tensor");
+  TORCH_CHECK(src_nodes.is_cpu(), "'src_nodes' must be a CPU tensor");
+  TORCH_CHECK(dst_nodes.is_cpu(), "'dst_nodes' must be a CPU tensor");
+
+  const auto num_nodes = rowptr.size(0) - 1;
+  at::Tensor out_rowptr, out_col;
+  c10::optional<at::Tensor> out_edge_id;
+
+  AT_DISPATCH_INTEGRAL_TYPES(
+      src_nodes.scalar_type(), "subgraph_bipartite_kernel", [&] {
+        // TODO: at::max parallel but still a little expensive
+        Mapper<scalar_t> mapper(at::max(col).item<scalar_t>() + 1,
+                                dst_nodes.size(0));
+        mapper.fill(dst_nodes);
+
+        auto res = subgraph_with_mapper<scalar_t>(rowptr, col, src_nodes,
+                                                  mapper, return_edge_id);
+        out_rowptr = std::get<0>(res);
+        out_col = std::get<1>(res);
+        out_edge_id = std::get<2>(res);
+      });
+
+  return {out_rowptr, out_col, out_edge_id};
+}
+
+std::tuple<at::Tensor, at::Tensor, c10::optional<at::Tensor>> subgraph_kernel(
+    const at::Tensor& rowptr,
+    const at::Tensor& col,
+    const at::Tensor& nodes,
+    const bool return_edge_id) {
+  return subgraph_bipartite_kernel(rowptr, col, nodes, nodes, return_edge_id);
+}
+
 }  // namespace
 
 TORCH_LIBRARY_IMPL(pyg, CPU, m) {
   m.impl(TORCH_SELECTIVE_NAME("pyg::subgraph"), TORCH_FN(subgraph_kernel));
+  m.impl(TORCH_SELECTIVE_NAME("pyg::subgraph_bipartite"),
```

Comment:
Any reason we want to expose that? Looks more like an internal function to me.

Reply:
If the user want to build a subgraph of a bipartite graph then he can use it.